### ISSUE TITLE
Investigate high loss of Mixtral

### ISSUE THREAD

**Issue by casper-hansen** [axolotl-ai-cloud/axolotl#931]
### ⚠️ Please check that this feature request hasn't been suggested before.

- [X] I searched previous [Ideas in Discussions](https://github.com/OpenAccess-AI-Collective/axolotl/discussions/categories/ideas) didn't find any similar feature requests.
- [X] I searched previous [Issues](https://github.com/OpenAccess-AI-Collective/axolotl/labels/enhancement) didn't find any similar feature requests.

### 🔖 Feature description

The axolotl implementation is not aligned with the [MegaBlocks implementation](https://github.com/stanford-futuredata/megablocks/blob/main/megablocks/layers/router.py#L56-L62).

```python
class LearnedRouter(torch.nn.Module):
    def forward(self, x):
        if self.training and self.args.moe_jitter_eps is not None:
            x = x * self.jitter(x)

        scores = self.layer(x.view(-1, x.shape[-1])).softmax(dim=-1)
        expert_weights, expert_indices = self._top_k(scores)

        expert_indices = (
            _uniform_expert_assignment(expert_indices, self.args.moe_num_experts)
            if self.args.uniform_expert_assignment else expert_indices
        )
        return scores, expert_weights, expert_indices
```

The current implementation:

https://github.com/OpenAccess-AI-Collective/axolotl/blob/68b227a7d8045d0f428d7ca3b9750f837d03611f/src/axolotl/models/mixtral/modeling_moe_mistral.py#L223-L232

Mistral also commented on this:

![image](https://github.com/OpenAccess-AI-Collective/axolotl/assets/27340033/dc8cead9-19e4-4c66-88e5-8968ab0c8164)

### ✔️ Solution

Investigate how we can adapt the most correct solution for the router. One way to test this is to measure the initial loss. For reference, back when I implemented sliding windows for Mistral, the initial loss dropped from 9.98 on main to 1.9 from the PR.

Measure loss on short and long context data, e.g. use `casperhansen/longalpaca_1k_test` with alpaca format.

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this feature has not been requested yet.
- [X] I have provided enough information for the maintainers to understand and evaluate this request.
